[1단계 - 출석] 헤일러(민서현) 미션 제출합니다.

threepebbles opened the pull request:
안녕하세요 샐리, 헤일러입니다 😃 
이번 미션 잘 부탁드립니다!

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [헤일러의 프롤로그](https://prolog.techcourse.co.kr/studylogs/3903)에 셀프 체크를 작성했나요?

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
step1에서는 처음으로 사용해보는 TDD 사이클에 익숙해지기, TDD 이후 구현된 기능들을 설계한 MVC 패턴에 맞게 조립하는 데에 집중하느라 객체지향 생활체조 요구사항을 잘 지키지 못한 것 같습니다.
특히 메서드에 한 단계의 들여쓰기, 한 줄에 점 하나만 찍기 요구사항을 지키지 못했습니다.

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
객체지향 생활체조 요구사항까지 지키면서 설계하지 못한 것이 아쉬웠습니다.

## 어떤 부분에 집중하여 리뷰해야 할까요?
이번 페어 프로그래밍을 진행하며 고민되었던 부분에 대해 공유드리고 싶습니다.
### 1️⃣ dto 패키지 위치에 대한 고민
view에서 하나의 기능 단위(ex. 출석 확인 기능, 출석 확인 기능)로 오고가는 데이터들을 dto로 묶어서 전송했습니다.
이번 프로젝트에서는 view <-> controller 간 사용하는 dto 클래스들을 {가장 상위 패키지/dto}에 위치하도록 했는데요.
현재 코드를 조금 더 리팩토링을 진행한다면 controller <-> domain간 dto를 사용할 수도 있을 수도 있다고 생각합니다. (step2에서 진행해보고 싶은 내용)
dto를 어느 패키지 위치에 두는 방식을 선호하시는지 샐리의 경험을 물어보고 싶습니다.

감사합니다 😄

songsunkook committed:
docs: 요구사항 명세서 작성

songsunkook committed:
red: 닉네임과 등교 시간으로 출석한다

songsunkook committed:
green: 닉네임과 등교 시간으로 출석한다

songsunkook committed:
refactor: 닉네임과 등교 시간으로 출석한다

songsunkook committed:
red: 이미 출석한 경우 예외를 던진다

songsunkook committed:
green: 이미 출석한 경우 예외를 던진다

songsunkook committed:
refactor: 닉네임과 등교 시간으로 출석한다

songsunkook committed:
red: 닉네임과 수정날짜와 등교시간으로 기록을 수정한다

songsunkook committed:
green: 닉네임과 수정날짜와 등교시간으로 기록을 수정한다

songsunkook committed:
red: 날짜와 시간으로 출석 상태를 계산한다

songsunkook committed:
green: 날짜와 시간으로 출석 상태를 계산한다

songsunkook committed:
refactor: 날짜와 시간으로 출석 상태를 계산한다

songsunkook committed:
docs: 요구사항 정리 및 입출력 예시 추가

songsunkook committed:
red: 출석 상태별 횟수를 계산한다

songsunkook committed:
green: 출석 상태별 횟수를 계산한다

songsunkook committed:
red: 관리 대상자 여부를 계산한다

songsunkook committed:
green: 관리 대상자 여부를 계산한다

songsunkook committed:
refactor: 관리 대상자 여부를 계산한다

songsunkook committed:
red: 출석 기록을 정상적으로 파싱한다

songsunkook committed:
green: 출석 기록을 정상적으로 파싱한다

songsunkook committed:
refactor: 출석 기록을 정상적으로 파싱한다

songsunkook committed:
docs: 기타 요구사항 추가 (TDD 종료)

songsunkook committed:
refactor: 패키지 구조 변경

songsunkook committed:
feat: 출석 확인 입출력 구현

songsunkook committed:
feat: 파일 입력 기반 초기 데이터 삽입 구현

songsunkook committed:
feat: 출석 수정 입출력 구현

songsunkook committed:
feat: 크루별 출석 기록 확인 입출력 구현

songsunkook committed:
feat: 제적 위험자 확인 입출력 구현

songsunkook committed:
feat: 캠퍼스 운영시간이 아닌 경우 예외 처리

songsunkook committed:
feat: 특정 기능 수행 후 초기 화면으로 복귀

songsunkook committed:
feat: 예외상황 처리

songsunkook committed:
refactor: 정렬 기준 수정

songsunkook committed:
fix: Crew.get() 버그 수정

songsunkook committed:
refactor: controller 메서드 분리

songsunkook committed:
refactor: else 제거

songsunkook committed:
feat: 공휴일 제외 기능 구현

songsunkook committed:
test: 테스트 추가 작성 및 리팩터링

threepebbles committed:
docs: 문자 수정(~ -> -)

threepebbles committed:
fix: 캠퍼스 운영시간 상수 수정

threepebbles committed:
refactor: 사용하지 않는 catch 구문 제거

threepebbles committed:
fix: 누락된 출력 요구사항 추가

dusdn1702 commented on src/main/java/dto/ModifiedResult.java:
```diff
@@ -0,0 +1,20 @@
+    InnerStatus after
+){
+
+    public record InnerStatus(
```
InnerStatus의 의미가 무엇인가요..? 🤔

threepebbles replied:
출석 시간과 출석 상태를 하나의 그룹으로 묶어서 표현하고 싶었는데 마땅한 네이밍이 생각나지 않았습니다..!

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,44 @@
+import java.util.Comparator;
+
+public enum AttendanceStatus {
+    ATTENDANCE("출석", Integer.MIN_VALUE),
```
reference type인 Integer 값을 넣어주신 이유가 무엇인가요?

threepebbles replied:
가장 작은 정수 값과 큰 정수 값 상수를 사용하고 싶어서 Integer에 정의된 값을 사용했습니다.

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,44 @@
+    ATTENDANCE("출석", Integer.MIN_VALUE),
+    LATE("지각", 6),
+    ABSENT("결석", 31),
+    NONE("", Integer.MAX_VALUE),
```
NONE은 어떤 상태인가요? 🤔

threepebbles replied:
출석, 지각, 결석을 정의할 수 없는 날에 대한 상태(주말 혹은 공휴일)를 정의하려고 했습니다!
description에 설명 추가했습니다.

dusdn1702 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,44 @@
+        long difference = MINUTES.between(lectureTime.getStartTime(), attendanceTime);
+        return Arrays.stream(values()).filter(attendanceStatus -> attendanceStatus.threshold <= difference)
+            .max(Comparator.comparing(AttendanceStatus::getThreshold))
+            .orElse(AttendanceStatus.NONE);
```
코드를 읽으면서 출석상태를 매핑하는데 threshold는 어떤 값이고, 비교해서 NONE으로 매핑하고 하는 과정들에서 복잡도가 높아 이해가 쉽지 않았던 것 같아요. 함수 분리를 통해 가독성을 높여보면 어떨까요?

threepebbles replied:
threshold, difference라는 단어로 의미전달이 되기 어려울 것이라 생각이 들어 elapsedMinutes(경과한 시간(단위: 분))를 사용하도록 수정했습니다. 
메서드 분리를 하기에는 작은 기능이라 생각해서 네이밍만 좀 더 명확하게 해주었는데 이러면 가독성이 높아졌을까요?
```java
public static AttendanceStatus of(LocalDate date, LocalTime attendanceTime) {
    LectureTime lectureTime = LectureTime.from(date);
    long elapsedMinutes = MINUTES.between(lectureTime.getStartTime(), attendanceTime);
    return Arrays.stream(values())
            .filter(attendanceStatus -> attendanceStatus.elapsedMinutesLimit < elapsedMinutes)
            .max(Comparator.comparing(AttendanceStatus::getElapsedMinutesLimit))
            .orElse(AttendanceStatus.NONE);
}
```

dusdn1702 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,90 @@
+
+    public Crew(String name) {
+        this.nickname = name;
+    }
```
일급컬렉션으로 객체 분리될 수 있는 여지가 많아보이는데 분리해보면 어떨까요?

threepebbles replied:
`Map<LocalDate, LocalTime> attendanceTimes` 을 일급컬렉션으로 분리하는 것을 말씀하시는건가요? 
단순히 어떤 날짜(LocalDate)에 어떤 시간(LocalTime)에 출석했다라는 의미를 가진 컬렉션인데 일급컬렉션을 사용함으로써 얻을 수 있는 장점이 있을까요?

dusdn1702 replied:
어떤 날짜(LocalDate)에 어떤 시간(LocalTime)에 출석했다라는 의미를 가졌다는 것만으로 객체로 분리할 이유가 되지 않을까요?
출석 정보를 변경하고 추가하고 삭제하고, 그 객체가 가질 책임이 많을 것 같은데요!
헤일러가 객체로 분리하는 기준은 무엇인가요?

dusdn1702 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,90 @@
+        this.nickname = name;
+    }
+
+    public AttendanceStatus attendance(LocalDate date, LocalTime time) {
```
메서드명을 보고 어떤 일을 수행하는지 알기 쉬우면 좋을 것 같아요!

threepebbles replied:
AttendanceTime(출석 시간)을 추가(기록)한다는 의미가 담기도록 수정했습니다!

dusdn1702 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public static void printHistory(AttendanceHistoryResult historyResult) {
+        System.out.printf("이번 달 %s의 출석 기록입니다.%n%n", historyResult.nickname());
```
출석 기록이 잘 나오지 않는 것 같아요.

threepebbles replied:
기록이 담긴 문자열을 마지막에 출력 안했네요! 수정했습니다.

dusdn1702 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,65 @@
+
+    public static OptionRequest scanOption() {
+        return RetryHandler.retryUntilSuccessWithReturn(() -> {
+            LocalDate now = DateTimeUtil.nowDate();
```
오늘로 시작이 되어서 정상적으로 수행되지 않는 기능들이 있는 것 같은데요~ 확인 부탁드릴게요!

dusdn1702 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,111 @@
+                case "3" -> checkAttendanceHistory();
+                case "4" -> checkCrewsAlmostExpelled();
+                case "q", "Q" -> isRunning = false;
+                default -> throw new IllegalArgumentException("존재하지 않는 기능입니다.");
```
오입력 시 에러로 실행 종료 보다는 에러 메시지 노출과 함께 재시도 할 수 있도록 하면 어떨까요?

threepebbles replied:
메인 화면에서 오입력 시 재입력 받도록 기능 추가했습니다!

dusdn1702 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    public void modifyAttendance(LocalDate date, LocalTime time) {
+        attendanceTimes.put(date, time);
```
같은 값을 넣어도 수정되는 경험이 사용자에게 좋을까요?

threepebbles replied:
기존에 기록된 출석 시간(HH:mm)과 수정하려는 시간(HH:mm)에 대한 비교하여 두 시간이 동일할 시 "수정하려는 시간이 기존 출석 시간과 동일합니다" 메세지를 띄워주고 수정 요청을 반려시키는 것을 말씀하시는 걸까요?

dusdn1702 replied:
네! 필수 조건은 아니지만 제안해봤어요 😄

dusdn1702 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,90 @@
+        return AttendanceStatus.of(date, attendanceTimes.get(date));
+    }
+
+    public List<History> getAllHistory(LocalDate today) {
```
List<History>, Map<AttendanceStatus, Integer> 모두 일급컬렉션 객체 분리가 가능해 보이네요!

dusdn1702 commented on src/main/java/domain/CrewRepository.java:
```diff
@@ -0,0 +1,57 @@
+
+import util.FileParser;
+
+public class CrewRepository {
+
+    private final List<Crew> crews = new ArrayList<>();
+
+    public CrewRepository(boolean initiailize) {
```
Repository의 역할을 무엇으로 생성해주신건지 궁금해요! 
initialize 여부로 생성 되는게 어색하게 느껴지네요 👀

threepebbles replied:
저도 CrewRepository는 단순히 CRUD기능만 수행하도록 하고 싶은데 initialize하는 책임도 가진 것이 어색하다 생각이 들었습니다.
그래서 출석 기록이 담긴 csv를 파싱하여 데이터를 추가하는 책임을 CrewRepository가 아닌 Controller에게 부여하도록 수정했습니다.
```java
public class ResourceLoader {
    public static void loadCrewRepository() {
        FileParser.loadAttendanceRecords()
                .forEach(record -> addAttendanceTime(
                        record.nickname(),
                        record.date(),
                        record.time()
                ));
    }
    ...
```
```java
public class AttendanceController {
    public AttendanceController() {
        ResourceLoader.loadCrewRepository();
    }
    ...
```

dusdn1702 replied:
좋네요! 💯

dusdn1702 commented on src/main/java/domain/Holiday.java:
```diff
@@ -0,0 +1,33 @@
+        this.day = day;
+    }
+
+    public int getMonth() {
+        return month;
+    }
+
+    public int getDay() {
+        return day;
+    }
```
사용처가 없는 것 같은데 존재해야하는 이유가 있을까요?

threepebbles replied:
초창기 코드에서 사용하다가 from을 구현하면서 사용하지 않게 된 함수네요! 삭제했습니다.

dusdn1702 commented on src/main/java/dto/AttendanceHistoryResult.java:
```diff
@@ -0,0 +1,42 @@
+    public static AttendanceHistoryResult of(String nickname, List<History> history, Map<AttendanceStatus,
+        Integer> statusCounter, Manage manage) {
+        List<InnerHistory> innerHistory = new ArrayList<>();
+        history.forEach(h -> {
```
의미있는 변수명을 사용해보면 어떨까요?

dusdn1702 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,90 @@
+            }
+            LocalDate date = today.withDayOfMonth(day);
+            LocalTime time = attendanceTimes.get(date);
+            histories.add(new History(date, time, getAttendanceStatusByDate(date), time == null));
```
값이 있는지 없는지 time == null 보다 더 좋은 방법을 사용할 수 있을 것 같아요!

dusdn1702 commented on src/main/java/domain/History.java:
```diff
@@ -0,0 +1,13 @@
+    LocalDate date,
+    LocalTime time,
+    AttendanceStatus status,
+    boolean isChecked
```
isChecked 의 의미가 출석인 것 같은데 변수명을 더 이해되기 쉽게 써보면 어떨까요?

threepebbles replied:
isChecked를 사용하지 않고 더 직관적이게 "30분 초과 결석"과 "출석 기록이 아예 되지 않은 결석"을 구분하기 위해 ABSENT_LATE, LATE, OFF_DAY을 추가했습니다!
```java
public enum AttendanceStatus {
    ATTENDANCE("출석", "출석 시간 준수", Integer.MIN_VALUE),
    LATE("지각", "출석 시간 5분 초과해서 지각", 5),
    ABSENT_LATE("결석", "출석 시간 30분 초과해서 결석", 30),
    ABSENT("결석", "출석 기록을 하지 않아서 결석", Integer.MAX_VALUE),
    OFF_DAY("쉬는 날", "출석 상태를 정의할 수 없음", Integer.MAX_VALUE),
    ;

    private final String title;
    private final String description;
    private final int elapsedMinutesLimit;
```

dusdn1702 commented on src/main/java/domain/Holiday.java:
```diff
@@ -0,0 +1,33 @@
+
+public enum Holiday {
+    CHRISTMAS(12, 25),
+    NONE(0, 0),
```
enum에서 NONE이 있는 경우와 없는 경우의 기준이 궁금해요!

threepebbles replied:
Holiday.NONE은 공휴일이 아닌 경우를 의미하도록 하려고 했는데 설명이 없어서 헷갈릴 수 있겠네요.
```java
public enum Holiday {
    CHRISTMAS("크리스마스", 12, 25),
    NOT_HOLIDAY("공휴일이 아닌 경우", 0, 0),
    ;

    private final String description;
    private final int month;
    private final int day;

    ...
```
NOT_HOLIDAY로 수정하고 description 추가 했습니다!

dusdn1702 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    private void checkCrewsAlmostExpelled() {
+        List<Crew> crews = crewRepository.getAll();
+        List<CrewAlmostExpelledResult> result = crews.stream()
+            .map(crew -> {
+                Map<AttendanceStatus, Integer> statusCounter
+                    = crew.getAttendanceStatusCounter(DateTimeUtil.nowDate());
```
객체 분리로 더욱 가독성 좋은 코드가 될 수 있을 것 같아요!

threepebbles replied:
`Map<AttendanceStatus, Integer> statusCounter`를 `AttendanceStatusStatistics`로 분리해보았습니다!

dusdn1702 commented on src/main/java/util/RetryHandler.java:
```diff
@@ -0,0 +1,15 @@
+
+import java.util.function.Supplier;
+
+public class RetryHandler {
```
요게 뭔가 의도하신대로 구현이 안되는 것 같네요~!

dusdn1702 commented on src/test/java/domain/CrewRepositoryTest.java:
```diff
@@ -0,0 +1,24 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class CrewRepositoryTest {
```
더 많은 테스트가 필요할 것 같네요!

dusdn1702 commented on src/test/java/domain/CrewTest.java:
```diff
@@ -0,0 +1,96 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+public class CrewTest {
```
전체적으로 규칙이 변경되어도 테스트 수정 필요 없이 로직만 검증할 수 있는 테스트로 리팩터링해보면 어떨까요?
예를 들어, 출석 기준 시간이 12시로 변경되면 모든 변수들을 찾아 수정하는 것은 어렵고 불필요한 작업 아닐까요?

threepebbles replied:
```java
public static final LocalDate MONDAY = LocalDate.of(2025, 2, 3);
public static final LocalDateTime MONDAY_ATTENDANCE = LocalDateTime.of(MONDAY,
        LectureTime.MONDAY
                .getStartTime()
                .plusMinutes(AttendanceStatus.LATE.getElapsedMinutesLimit()));
public static final LocalDateTime MONDAY_LATE = LocalDateTime.of(MONDAY,
        LectureTime.MONDAY
                .getStartTime()
                .plusMinutes(AttendanceStatus.LATE.getElapsedMinutesLimit() + 1));
public static final LocalDateTime MONDAY_ABSENT_LATE = LocalDateTime.of(MONDAY,
        LectureTime.MONDAY
                .getStartTime()
                .plusMinutes(AttendanceStatus.ABSENT_LATE.getElapsedMinutesLimit() + 1));
```
위와 같이 기존에 정의해 놓은 상수를 이용해서 테스트 픽스처를 만들어놓고 사용하면 해결할 수 있는 문제일까요?

dusdn1702 replied:
네! 저라면 테스트 변수들을 이런 값을 이용했을 것 같아요 👍

dusdn1702 commented on src/test/java/domain/ManageTest.java:
```diff
@@ -0,0 +1,66 @@
+public class ManageTest {
+
+    @Test
+    void 결석_2회_미만인_경우_아무_관리_대상자가_아니다() {
```
어디선 메서드명에 테스트 설명을, 어디선 displayName을 사용하고 계신데 통일하면 좋을 것 같아요!

threepebbles replied:
DisplayName에서 한글 설명을 사용하도록 일관되게 수정했습니다!

dusdn1702 reviewed (changes requested):
안녕하세요 헤일러!
짧은 시간에 많은 고민을 하고 구현해주시느라 고생 많으셨습니다. 

dto에 대한 말씀을 주셨는데 dto가 필요한 이유가 뭐라고 생각하시나요? 오히려 불필요한 매핑로직이 생기는건 아닐까요? 말씀 주신 것처럼 controller <-> domain 간 dto가 추가되면, 같은 값이더라도 새로운 dto 클래스를 생성하는 것이 좋을까요? 생성하고 사용하는 이유가 명확해지면 패키지 위치에 대한 고민도 해소될 것 같아요 😄 

그 외에 리뷰하면서 궁금한 점들 코멘트 남겨두었습니다. 확인 부탁드릴게요 !
또 궁금한 것들도 편하게 댓글 남겨주시면 같이 확인하도록 하겠습니다. 🙇

threepebbles committed:
refactor: CrewRepository 내 메서드를 모두 정적 메서드로 변경

threepebbles committed:
fix: 메인 화면에서 사용자가 잘못된 입력을 한 경우 에러메세지 출력 후 재입력 받도록 수정

threepebbles committed:
refactor: 메서드명 변경

threepebbles committed:
refactor(AttendanceStatus): 변수 네이밍 변경

threepebbles committed:
feat(Crew): 출석 기록이 존재하지 않는 경우 예외가 발생하도록 수정

threepebbles committed:
feat(AttendanceStatus): 결석을 두 종류로 분류(출석 기록이 늦은 결석, 오지 않은 결석

threepebbles committed:
refactor(AttendanceStatusStatistice): 일급 컬렉션 추가 및 적용

threepebbles committed:
fix: 메인 화면에서 올바르지 않은 옵션 입력 시 재입력 받도록 수정

threepebbles committed:
refactor: 예외 메시지 변경

threepebbles committed:
refactor: 조건문 정리

threepebbles committed:
refactor(MonthAttendanceRecordsResult): 필요 없는 메서드 제거

threepebbles committed:
refactor: 공백 제거

threepebbles committed:
refactor: 사용하지 않는 조건문 제거

threepebbles committed:
refactor(Formatter): 패키지 위지 변경(dto -> util)

threepebbles committed:
refactor(Holiday): description 필드 추가

threepebbles committed:
refactor(LectureTime): 교육시간이 아닌 상수(SATURDAY, SUNDAY) 제거

threepebbles committed:
refactor: 네이밍 변경

threepebbles committed:
refactor(CrewRepository): 패키치 위치 이동

threepebbles committed:
fix: 사용자 입력에 대해 예외 발생 시 재입력 받도록 기능 추가

threepebbles committed:
refactor: 네이밍 컨벤션에 맞게 변경

threepebbles committed:
fix: 누락된 출석 기록 출력 기능 추가

threepebbles committed:
test: CrewRepository 기능 테스트 추가

threepebbles committed:
refactor: 메서드 분리

threepebbles committed:
refactor: 클래스 및 메서드 분리

threepebbles committed:
test: 테스트 실행마다 다른 given이 주어지는 테스트 수정

threepebbles committed:
refactor: 한달 출석 기록을 추출하는 기능의 메서드 위치 이동

threepebbles committed:
refactor(AttendanceTime): AttendanceTime 객체 분리

threepebbles committed:
feat: service 기능 실행 중 Exception 발생 시 프로그램을 종료하지 않고 예외 메시지만 출력하도록 수정

threepebbles committed:
refactor: 예외 메시지 수정

threepebbles committed:
refactor: 예외 메시지 수정

threepebbles committed:
docs: 비즈니스 로직과 입출력 요구 사항 분리하여 정리

threepebbles committed:
refactor: 예외 메시지 수정

threepebbles committed:
refactor: 예외 메시지 수정

threepebbles committed:
docs: 누락된 요구 사항 추가

threepebbles committed:
refactor: 크루별 출석 기록 출력 기능 메서드 분리

threepebbles committed:
perf: 정렬 기능 시간 복잡도 개선

threepebbles committed:
refactor: 가독성을 높이기 위해 for loop 대신 stream 사용

threepebbles committed:
feat(Crew): 해당 날짜(월일)에 출석 기록이 존재하지 않는 경우, 예외가 아닌 의미 있는 객체를 반환하도록 수정

threepebbles committed:
feat: 필요 없는 조건문 제거

threepebbles committed:
fix: 출력할 리스트 수정

threepebbles committed:
fix: 제적 위험자 목록 정렬 기준 수정(지각+결석 오름차순 -> 내림차순)

threepebbles committed:
refactor: 메서드 분리

threepebbles committed:
chore: 출석 기록 csv 파일 수정

threepebbles committed:
refactor: StringBuilder -> String 수정

String의 += 연산을 사용하는 것이 아닌 +로 한번에 이어붙이는 경우 컴파일러에서 자동으로 StringBuilder를 사용하도록 최적화함. 동일한 성능 보장.

threepebbles committed:
fix: 누락되었던 주말 및 공휴일을 검증 기능 추가

threepebbles committed:
chore: 테스트에서 사용되는 닉네임에 대한 기록 삭제

threepebbles committed:
comment: TODO 추가

threepebbles committed:
feat(AttendanceTime): null 방지 조건 추가

threepebbles reviewed (commented):
안녕하세요 샐리!
피드백 해주신 내용 기반으로 리팩토링을 진행했습니다.
특히 이전 코드에 비해 가독성을 높이기 위해 많은 노력을 했습니다.
이번에도 리뷰 잘 부탁드립니다.
감사합니다 😄

## 주요 변경 사항

- 버그 fix
    - 테스트 코드에서 now를 사용하여 실행마다 given이 달라지던 버그
    - 예외 조건에서 프로그램이 예외와 함께 종료되는 버그
    - 등등 많은 자잘한 버그들..
- 가독성을 높이기 위한 노력
    - 의미있는 변수 네이밍
    - 일급 컬렉션, 객체 분리
        - ex)
            - AttendanceRecord: 어떤 날짜(월일)에 대한 출석 기록
            - AttendanceTime: 출석 시간(시분)과 출석 상태
            - AttendanceStatus: 출석 상태(출석, 지각, 30분 초과한 결석, 출석 기록이 없는 결석)
            - AttendanceStatusStatistics: 출석 상태 별 횟수
- 서비스, 레포지토리 계층 분리 및 리팩터링
    - 도메인에서 생성된 객체가 뷰영역까지 그대로 내려와 사용되는 기능이 있어서 패키지 별로 관심사를 분리하고 의존관계를 약화시키기 위해 서비스와 레포지토리 계층을 추가적으로 분리해보았습니다.

## 질문 드리고 싶은 내용

### :one: null을 사용하지 않고 설계하는 방법

이번 미션에서 _크루별 출석 기록 확인_ 기능을 구현할 때 null을 사용하지 않고 설계하는 것이 어려웠습니다.

리뷰 첫 라운드 이전에는 Crew 클래스 안에서 Map 자료구조(`Map<LocalDate, LocalTime> attendanceTimes`)를 이용해서 특정 날짜(LocalDate)에 기록된 출석 시간(LocalTime)을 가져오도록 구현했었는데요.

이 자료구조를 이용해 원하는 날짜에 대한 출석 시간 정보를 가져오는 메서드를 null을 사용하지 않고 구현하는 방법이 도저히 생각나지 않았습니다.

그래서 아래와 같이 AttendanceTime이라는 객체를 분리했고, 항상 AttendanceTime의 AttendanceStatus를 먼저 검증함으로써 LocalTime에 대한 null 검증을 할 일이 없게 설계했습니다.

이런 방식의 설계가 유효한 방법인지, 허점은 없는지, 혹시 다른 더 좋은 설계 방법이 있을지 여쭤보고 싶습니다!

```java
// AttendanceTime.java
public record AttendanceTime(
        LocalTime time,
        AttendanceStatus status
) {
    public AttendanceTime {
        if (status == null) {
            throw new IllegalArgumentException("출석 상태는 null일 수 없습니다.");
        }
    }
}

```

```java
// Crew.java
public class Crew {
    private final String nickname;
    private final Map<LocalDate, AttendanceTime> attendanceTimes = new HashMap<>();
    ...

    public AttendanceTime getAttendanceTimeByDate(LocalDate date) {
        if (!attendanceTimeExists(date)) {
            return new AttendanceTime(null, AttendanceStatus.ABSENT);
        }
        return attendanceTimes.get(date);
    }
    ...
}
```

dusdn1702 commented on src/main/java/view/OutputView.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    public static void printCrewsAlmostExpelled(List<CrewAlmostExpelledResult> result) {
+        // stream.sort()는 Arrays.sort()와 동일. 퀵소트 기반이므로 최악에 O(n^2)
+//        result = result.stream()
+//                .sorted(
+//                        Comparator.comparing(CrewAlmostExpelledResult::calculateTotalAbsentCount,
+//                                        Comparator.reverseOrder())
+//                                .thenComparing(CrewAlmostExpelledResult::nickname))
+//                .toList();
+        // Collections.sort()는 머지소트 기반이므로 O(nLog(n)) 보장
```
처리 안된 주석이 있는 것 같네요~! 👀

dusdn1702 commented on src/main/java/util/AttendanceRecord.java:
```diff
@@ -0,0 +1,11 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+
+public record AttendanceRecord(
```
요 값은 어떤 역할일까요? 
단순히 파싱만을 위한 값이라면 객체명이 바뀌어야하지 않을까요?
dto 패키지에 두지 않으신 이유도 궁금해요!

threepebbles replied:
기존 코드에서 Crew 객체가 AttendanceRecord에 의존하고 있었는데 메서드 분리를 하는 과정에서 의존하지 않게 구조가 바뀌었어요. 분리한 이유가 AttendanceRecord를 dto로 분리하기 위해서였는데 패키지 옮기는 것을 깜빡했네요 🥹 감사합니다.

dusdn1702 reviewed (approved):
안녕하세요 헤일러 !
피드백 반영해주신 부분들 확인했습니다. 의존관계까지 고려해주신 부분이 인상 깊었어요 💯 

1️⃣ -> 저는 이렇게 객체 분리를 통해 책임을 나눔으로써 말씀 주신 null 제어 뿐만 아니라 추후 유지보수를 위한 리팩터링 측면에서도 큰 장점이 있다고 생각해요! 👍

이번 단계는 여기서 마무리 짓고 남겨둔 몇가지 코멘트는 다음 단계에서 확인 부탁드릴게요~
고생 많으셨습니다. 다음 단계도 화이팅입니다! 🚀
